### PR TITLE
DDP bug fix

### PR DESCRIPTION
**Pull request by pzhanggit**
Major features/changes:
1. Fixed a bug in DDP model. The DDP wrapped model was unwrapped (which is not supposed to be) before training. As a result, every process trains its own model with a subset of dataset and there is no synchronization between them. 
2. Fixed a bug about map_location in model loading
3. Added reduce/gather feature for loss log and results collection
4. Added SyncBatchNorm 
MAE_GPUs_comp_test_ddpbug

Checked that the gradients and model parameters on rank 0 and rank 1 are the same
gradient_check
parameters_check

### CODE REVIEW

**Comment by allaffa** [comment]
@pzhanggit 
I am running the code on FePt. 
I will let you know what I obtain as a result. 
While I am at it, do you want me to run the PR on other datasets that you did not run yourself?

**Comment by pzhanggit** [comment]
> @pzhanggit I am running the code on FePt. I will let you know what I obtain as a result. While I am at it, do you want me to run the PR on other datasets that you did not run yourself?

Thanks, @allaffa. What datasets you have in mind? So far all I ran is the dftb version of qm9 and in the single head mode. Could you run one case with multihead?

**Comment by jychoi-hpc** [comment]
This is good. I will run a test on Summit. Which is the part doing "unwrapping" ddp model which is not supposed to happen?

File: hydragnn/train/train_validate_test.py [hunk 1]
```diff
@@ -46,16 +48,16 @@ def train_validate_test(
...
-    total_loss_test = [None] * num_epoch
+    total_loss_train = torch.zeros(num_epoch, device=device)
+    total_loss_val = torch.zeros(num_epoch, device=device)
+    total_loss_test = torch.zeros(num_epoch, device=device)
     # loss tracking for each head/task
-    task_loss_train = [None] * num_epoch
-    task_loss_test = [None] * num_epoch
-    task_loss_val = [None] * num_epoch
-
-    model = get_model_or_module(model)
```
Comment (pzhanggit) [inline hydragnn/train/train_validate_test.py:58]: @jychoi-hpc  the unwrapping here

**Comment by allaffa** [comment]
@pzhanggit 
Now we print the full pytorch tensor on the screen, which causes less digits to be printed out for the individual loss functions on each task. 
We should roll back to there same print formatting as before, maybe calling `torch.item()` instead of printing the entire tensor

File: hydragnn/utils/model.py [hunk 2]
```diff
@@ -75,7 +76,7 @@ def load_existing_model(model, model_name, path="./logs/", optimizer=None):
     """Load both model and optimizer state from a single checkpoint file"""
     _, world_rank = get_comm_size_and_rank()
     path_name = os.path.join(path, model_name, model_name + ".pk")
-    map_location = {"cuda:%d" % 0: "cuda:%d" % world_rank}
+    map_location = {"cuda:%d" % 0: get_device_name()}
```
Comment (streeve) [inline hydragnn/utils/model.py:79]: Maybe we rename these `get_local_device_name` and `get_global_size_and_rank`?

File: hydragnn/postprocess/visualizer.py [hunk 3]
```diff
@@ -282,6 +283,8 @@ def create_parity_plot_and_error_histogram_scalar(
     ):
         """Creates scatter plots for true_values and predicted values of variable varname at iepoch."""
 
+        true_values = true_values.cpu().detach().numpy()
```
Comment (streeve) [inline hydragnn/postprocess/visualizer.py:286]: Good idea

File: hydragnn/train/train_validate_test.py [hunk 4]
```diff
@@ -102,7 +105,11 @@ def train_validate_test(
             )
         val_loss, val_taskserr = validate(val_loader, model, verbosity)
         test_loss, test_taskserr, true_values, predicted_values = test(
-            test_loader, model, verbosity
+            test_loader,
+            model,
+            verbosity,
+            reduce_gpus=True,
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:111]: reduce_ddp? or reduce_ranks?

File: hydragnn/train/train_validate_test.py [hunk 5]
```diff
@@ -205,6 +218,67 @@ def get_head_indices(model, data):
...
+    if dist.get_world_size() > 1:
+        dist.all_reduce(total_error, op=dist.ReduceOp.SUM)
+        total_error = total_error / dist.get_world_size()
+        dist.all_reduce(tasks_error, op=dist.ReduceOp.SUM)
+        tasks_error = tasks_error / dist.get_world_size()
+    return total_error, tasks_error
+
+
+@torch.no_grad()
+def gather_head_values(head_values):
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:243]: gather_array_ddp?

File: hydragnn/train/train_validate_test.py [hunk 6]
```diff
@@ -205,6 +218,67 @@ def get_head_indices(model, data):
     return head_index
 
 
+@torch.no_grad()
+def reduce_loss_stat(total_loss_list, task_loss_list):
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:222]: Can we reuse one `reduce_values_ddp` function that takes a single list and returns the reduction or updates in place? Unless we really need both options

File: hydragnn/train/train_validate_test.py [hunk 7]
```diff
@@ -205,6 +218,67 @@ def get_head_indices(model, data):
...
+    # gather error statistics across all processes
+    if dist.get_world_size() > 1:
+        for total_error, tasks_error in zip(total_loss_list, task_loss_list):
+            dist.all_reduce(total_error, op=dist.ReduceOp.SUM)
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:226]: We discussed this a bit already, but I don't see a reason to all_gather instead of gather to rank zero (here and below)

File: hydragnn/utils/distributed.py [hunk 8]
```diff
@@ -213,7 +213,7 @@ def is_model_distributed(model):
     return isinstance(model, torch.nn.parallel.distributed.DistributedDataParallel)
 
 
-def get_distributed_model(model, verbosity=0):
+def get_distributed_model(model, verbosity=0, SyncBatchNorm=False):
```
Comment (streeve) [inline hydragnn/utils/distributed.py:216]: sync_batch_norm

File: hydragnn/models/Base.py [hunk 9]
```diff
@@ -94,7 +94,6 @@ def __init__(
         self._init_conv()
         if self.freeze_conv:
             self._freeze_conv()
-        self._init_node_conv()
```
Comment (streeve) [inline hydragnn/models/Base.py:97]: This seems unrelated

File: hydragnn/train/train_validate_test.py (hunk 1 above)
Comment (streeve) [reply hydragnn/train/train_validate_test.py:58]: We should remove that function entirely, right?

**Review by streeve — COMMENTED** [review]
This looks like a lot of work and I'm very glad the results make more sense once we're actually using DDP :)

**Review by allaffa — APPROVED** [review]

File: hydragnn/postprocess/visualizer.py [hunk 10]
```diff
@@ -116,9 +116,10 @@ def __scatter_impl(
         y_label=None,
         xylim_equal=False,
     ):
+
         ax.scatter(x, y, s=s, edgecolor="b", marker=marker, facecolor="none")
 
-        ax.set_title(title)
+        ax.set_title(title + ", len=" + str(len(x)))
```
Comment (allaffa) [inline hydragnn/postprocess/visualizer.py:122]:
```suggestion
        ax.set_title(title + ", number data samples = " + str(len(x)))
```

File: hydragnn/utils/distributed.py [hunk 11]
```diff
@@ -213,12 +213,27 @@ def is_model_distributed(model):
     return isinstance(model, torch.nn.parallel.distributed.DistributedDataParallel)
 
 
-def get_distributed_model(model, verbosity=0):
-    device_name = get_device_name(verbosity)
+def get_distributed_model(model, verbosity=0, SyncBatchNorm=False):
+    device_name = get_device_name(verbosity_level=verbosity)
+
+    world_size, world_rank = get_comm_size_and_rank()
+    print(
```
Comment (allaffa) [inline hydragnn/utils/distributed.py:220]: I would use print verbosity here. What do you think?

File: hydragnn/train/train_validate_test.py [hunk 12]
```diff
@@ -216,89 +290,97 @@ def train(
...
-        total_error += loss.item() * data.num_graphs
-        num_samples_local += data.num_graphs
-        for itask in range(len(tasks_loss)):
-            tasks_error[itask] += tasks_loss[itask].item() * data.num_graphs
-    return (
-        total_error / num_samples_local,
-        tasks_error / num_samples_local,
-    )
+        with torch.no_grad():
+            total_error += loss * data.num_graphs
```
Comment (allaffa) [inline hydragnn/train/train_validate_test.py:310]: any reason why you do not access the .item() anymore? 
The older printout to the screen looks better. More digits are printed on the screen with the old way

**Review by allaffa — CHANGES_REQUESTED** [review]

File: hydragnn/train/train_validate_test.py (hunk 7 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py:226]: Agreed. We will open an issue and make another PR as we discussed

File: hydragnn/models/Base.py (hunk 9 above)
Comment (pzhanggit) [reply hydragnn/models/Base.py:97]: It is relevant. Previously, conv for node is defined even it is not used in the model. However, DDP does not support unused parameters. So we change it to only define it when used.

File: hydragnn/utils/distributed.py (hunk 11 above)
Comment (pzhanggit) [reply hydragnn/utils/distributed.py:220]: Probably no difference here since we want to print on every rank

File: hydragnn/train/train_validate_test.py (hunk 12 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py:300]: Changed

File: hydragnn/train/train_validate_test.py (hunk 1 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py:58]: Not sure. It is used also in the model saving. Probably related to the non-DDP mode we were discussing earlier. Let's keep it for now.

File: hydragnn/utils/model.py (hunk 2 above)
Comment (pzhanggit) [reply hydragnn/utils/model.py:79]: Not relevant to this PR?

File: hydragnn/train/train_validate_test.py (hunk 4 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py]: Done

File: hydragnn/train/train_validate_test.py (hunk 5 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py]: Done

File: hydragnn/train/train_validate_test.py (hunk 6 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py]: Done

File: hydragnn/utils/distributed.py (hunk 8 above)
Comment (pzhanggit) [reply hydragnn/utils/distributed.py]: Done

**Review by allaffa — APPROVED** [review]

**Review by jychoi-hpc — APPROVED** [review]

File: hydragnn/models/Base.py [hunk 13]
```diff
@@ -341,6 +348,8 @@ def __init__(self, input_dim, output_dim, num_nodes, hidden_dim_node, node_type)
                 denselayers.append(ReLU())
             denselayers.append(Linear(hidden_dim_node[-1], output_dim))
             self.mlp.append(Sequential(*denselayers))
+            if self.node_type == "mlp":
```
Comment (streeve) [inline hydragnn/models/Base.py:351]: Let's switch to a variable loop bound instead

File: hydragnn/utils/distributed.py (hunk 11 above)
Comment (streeve) [reply hydragnn/utils/distributed.py:220]: `print_distributed` or remove

**Comment by pzhanggit** [comment]
> Should we remove `get_model_or_module`? It seems like a good way to make a mistake again

Not sure. It is used also in the model saving. Probably related to the non-DDP mode we were discussing earlier. Let's keep it for now.

File: hydragnn/utils/distributed.py (hunk 11 above)
Comment (pzhanggit) [reply hydragnn/utils/distributed.py]: Done in 882bac3e57f5cafc07c3f6a69cc7efab35ee9cc7

File: hydragnn/models/Base.py (hunk 13 above)
Comment (pzhanggit) [reply hydragnn/models/Base.py]: Done in 882bac3e57f5cafc07c3f6a69cc7efab35ee9cc7

File: hydragnn/train/train_validate_test.py [hunk 14]
```diff
@@ -216,89 +280,99 @@ def train(
...
+
     for data in iterate_tqdm(loader, verbosity):
         with record_function("zero_grad"):
             opt.zero_grad()
         with record_function("get_head_indices"):
             head_index = get_head_indices(model, data)
         with record_function("forward"):
             pred = model(data)
-            loss, tasks_loss = model.loss(pred, data.y, head_index)
+            loss, tasks_loss = model.module.loss(pred, data.y, head_index)
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:294]: I think this means we always require DDP from now on - the non-DDP doesn't have `module` defined, right?

**Review by streeve — APPROVED** [review]
Not sure if we want to default to True for the reductions, but we can always change it later if it's an issue

File: hydragnn/train/train_validate_test.py (hunk 14 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py:294]: Yes.
